components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,382 @@
+    private val statusServiceStub = StatusServiceGrpc.newStub(GitpodManager.supervisorChannel)
+    private val backendTerminalManager = BackendTerminalManager.getInstance(project)
+
+    init {
+        afterTerminalToolWindowGetsRegistered {
+            withSupervisorTasksList { tasksList ->
+                withSupervisorTerminalsList { terminalsList ->
+                    runInEdt {
+                        for (terminalWidget in terminalView.widgets) {
+                            val terminalContent = terminalView.toolWindow.contentManager.getContent(terminalWidget)
+                            val terminalTitle = terminalContent.getUserData(TITLE_KEY)
+                            if (terminalTitle != null) {
+                                debug("Closing terminal $terminalTitle before opening it again.")
+                                terminalWidget.close()
+                            }
+                        }
+
+                        if (tasksList.isEmpty() || terminalsList.isEmpty()) {
+                            backendTerminalManager.createNewSharedTerminal(
+                                    "GitpodTerminal",
+                                    "Terminal"
+                            )
+                        } else {
+                            val aliasToTerminalMap:
+                                    MutableMap<String, TerminalOuterClass.Terminal> =
+                                    mutableMapOf()
+
+                            for (terminal in terminalsList) {
+                                val terminalAlias = terminal.alias
+                                aliasToTerminalMap[terminalAlias] = terminal
+                            }
+
+                            for (task in tasksList) {
+                                val terminalAlias = task.terminal
+                                val terminal = aliasToTerminalMap[terminalAlias]
+
+                                if (terminal != null) {
+                                    createSharedTerminal(terminal)
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun afterTerminalToolWindowGetsRegistered(action: () -> Unit) {
+        debug("Waiting for TerminalToolWindow to be registered...")
+        val toolWindowManagerListener =
+                object : ToolWindowManagerListener {
+                    override fun toolWindowsRegistered(
+                            ids: MutableList<String>,
+                            toolWindowManager: ToolWindowManager
+                    ) {
+                        if (ids.contains(TerminalToolWindowFactory.TOOL_WINDOW_ID)) {
+                            debug("TerminalToolWindow got registered!")
+                            action()
+                        }
+                    }
+                }
+
+        project.messageBus
```

Comment:
Should not we check here first whether it is already registered?

Reply:
You mean `project`? If so, it's not needed, as the project is guaranteed to be registered at this point. Because we're calling this class as a `projectService`, it will only be initialized after the project is registered.

```xml
<projectService serviceImplementation="io.gitpod.jetbrains.remote.GitpodTerminalService" preload="true"/>
```

But if you're talking about `messageBus`, I'm really not checking it. I thought it was always set, for any project.